src/plugins/obj/obj_s3_client.cpp

```diff
@@ -0,0 +1,222 @@
+    if (virtual_addressing_it != custom_params->end()) {
+        const std::string &value = virtual_addressing_it->second;
+        if (value == "true")
+            return true;
+        else if (value == "false")
+            return false;
+        else
+            throw std::runtime_error ("Invalid value for use_virtual_addressing: '" + value +
+                                      "'. Must be 'true' or 'false'");
+    }
+
+    return false;
+}
+
+std::string
+getBucketName (nixl_b_params_t *custom_params) {
+    if (custom_params) {
+        auto bucket_it = custom_params->find ("bucket");
+        if (bucket_it != custom_params->end() && !bucket_it->second.empty()) {
+            return bucket_it->second;
+        }
+    }
+
+    const char *env_bucket = std::getenv ("AWS_DEFAULT_BUCKET");
+    if (env_bucket && env_bucket[0] != '\0') return std::string (env_bucket);
+    throw std::runtime_error ("Bucket name not found. Please provide 'bucket' in custom_params or "
+                              "set AWS_DEFAULT_BUCKET environment variable");
+}
+
+} // namespace
+
+AwsS3Client::AwsS3Client (nixl_b_params_t *custom_params,
+                          std::shared_ptr<Aws::Utils::Threading::Executor> executor)
+    : aws_options_ (
+          []() {
+              auto *opts = new Aws::SDKOptions();
+              Aws::InitAPI (*opts);
+              return opts;
+          }(),
+          [] (Aws::SDKOptions *opts) {
+              Aws::ShutdownAPI (*opts);
+              delete opts;
+          }) {
+    auto config = ::createClientConfiguration (custom_params);
+    if (executor) config.executor = executor;
+
+    auto credentials_opt = ::createAWSCredentials (custom_params);
+    bool use_virtual_addressing = ::getUseVirtualAddressing (custom_params);
+    bucket_name_ = Aws::String (::getBucketName (custom_params));
+
+    if (credentials_opt.has_value())
+        s3_client_ = std::make_unique<Aws::S3::S3Client> (
+            credentials_opt.value(),
+            config,
+            Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::RequestDependent,
+            use_virtual_addressing);
+    else
+        s3_client_ = std::make_unique<Aws::S3::S3Client> (
+            config,
+            Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::RequestDependent,
+            use_virtual_addressing);
+}
+
+void
+AwsS3Client::setExecutor (std::shared_ptr<Aws::Utils::Threading::Executor> executor) {
+    throw std::runtime_error ("AwsS3Client::setExecutor() not supported - AWS SDK doesn't allow "
+                              "changing executor after client creation");
+}
+
+void
+AwsS3Client::PutObjectAsync (std::string_view key,
+                             uintptr_t data_ptr,
+                             size_t data_len,
+                             size_t offset,
+                             PutObjectCallback callback) {
+    // AWS S3 doesn't support partial put operations with offset
+    if (offset != 0) {
```

Comment:
This check is only for Put and not Get, does that mean AWS supports partial read with get?